### ISSUE TITLE
Attacks by introducing invisible source code using special Unicode characters

### ISSUE THREAD

**Issue by MSoegtropIMC** [ocaml/ocaml#10749]
In this paper: (https://www.trojansource.codes/trojan-source.pdf) researchers from the University of Cambridge describe an attack which uses special Unicode characters to inject invisible or modified code. As far as I can tell any compiler taking Unicode input - as ocamlc - is affected by this.

There is an easy remedy, which is rejecting sources which contain such characters.

**Comment by nojb** [comment]
There is a discussion of this issue over at Discuss https://discuss.ocaml.org/t/is-ocaml-vulnerable-to-https-trojansource-codes/8723

**Comment by MSoegtropIMC** [comment]
Thanks for the pointer! I don't follow the argument given there that it is mostly an editor issue.

But an interesting question would be what github diffs make out of this, since this is the place where most code reviews happen.

**Comment by dbuenzli** [comment]
> But an interesting question would be what github diffs make out of this, since this is the place where most code reviews happen.

I made a gist with the example I gave in the discuss thread. The first commit has no vulnerability. Then I make the commit that introduces the vulnerability. 

The answer to the question is that you get a big warning from github that bidi characters are present in the file. On the diff and on the file itself. See for yourself here: 

https://gist.github.com/dbuenzli/e655590faf132d5150f3684a3d7a69a2/revisions

**Comment by MSoegtropIMC** [comment]
@dbuenzli : thanks for testing this on Github - we see that GitHub works pretty fast on this (as it looks this message was introduced October 31st).

But I don't think one can expect that in every environment code review processes are adjusted to this (and possibly upcoming similar issues), so IMHO the best strategy is to work on both, editors + review process tools and compilers.

Are there strong arguments against introducing a warning for this?

**Comment by lthls** [comment]
I fail to see any reason to get so excited about this particular issue.
Here is an adaptation of the example in the paper that causes the same issues, but using zero-width space:
https://gist.github.com/lthls/95e029513d6135c21c9fff09bcd4a3e2
(Note that while creating the gist, the Github interface clearly showed me the zero-width space, but it's invisible when rendered, even in the "raw" version).
Unicode is full of these little traps, and I don't think the compiler should try to catch any of them (certainly not all of them).

**Comment by Octachron** [comment]
Bidirectional formatting characters are a legitimate part of unicode. Emitting warning on those characters will lead to false positive. Moreover, since there is no constraint on the encoding of source code in OCaml, the compiler has no idea on how a string literal will be interpreted by graphical editors: will the graphical editor use utf-16, utf-8, another proprietary encoding?

Which leads to the core of the issue: the vulnerability is not in languages but purely in the potentially misleading graphical display of strings by editors. For instance, the vulnerability simply don't exist at all for screen reader users. For them, any warning on bidirectional formatting characters will have a 100% positive rate. Similarly, it doesn't make sense for build tools to emit such warning when there is no interactive users that could be mislead by potentially faulty string rendering.

> But I don't think one can expect that in every environment code review processes are adjusted to this (and possibly upcoming similar issues), 

This seems an inconsistent reaction to me: if you consider bidirectional formatting characters a meaningful attack vector, you ought to use editors that are not vulnerable to this attack vector.

I am still looking a sensible way to implement such a warning that doesn't trigger on all use of bidirectional formatting characters, but this still feel like protecting our users from faulty editors more than anything.

**Comment by dbuenzli** [comment]
> Unicode is full of these little traps, and I don't think the compiler should try to catch any of them (certainly not all of them).

This would need a bit more investigation but while Unicode may be full of these traps, they actually may be quite easy to characterize. I suspect they are mostly those characters whose general [general category](http://www.unicode.org/reports/tr44/#General_Category_Values) is `Cc` and `Cf`.

> Bidirectional formatting characters are a legitimate part of unicode. Emitting warning on those characters will lead to false positive.

The idea would be that writing them as escapes does not trigger the warning. Also I don't expect bidi control characters to be often present in string literals, false positives will likely be extremely rare. 

But the problem with warnings is that they can also be disabled and from the sources itself. I wonder if in the end these things would not be better checked by some form of external dedicated security linter tool.

**Comment by alainfrisch** [comment]
> But the problem with warnings is that they can also be disabled and from the sources itself. I wonder if in the end these things would not be better checked by some form of external dedicated security linter tool.

Well, if you assume people who submit code for inclusion in your project can be hostile, you need to be super careful about code review anyway.  The problem is code being hidden in a way which makes it hard to even see what the code actually is. Turning warnings off should be easy enough to spot...

> I suspect they are mostly those characters whose general general category is Cc and Cf.

If we have a clear criterion and an easy way of implementing it, I'm personally in favor of a new warning (enabled by default).  With the current feature freeze and the possibly delay in adopting 5.00, this won't be available before months in practice, though, unless we backport it to older versions.

**Comment by MSoegtropIMC** [comment]
> > But I don't think one can expect that in every environment code review processes are adjusted to this (and possibly upcoming similar issues),
> 
> This seems an inconsistent reaction to me: if you consider bidirectional formatting characters a meaningful attack vector, you ought to use editors that are not vulnerable to this attack vector.

In companies with a few 10.000 engineers one can't easily change the tools used in review processes from one day to another and one also cannot always harden them against such issues from one day to another. But one can quickly change compiler versions - they are anyway kept up to date for security reasons. It is new (to me) that one has to keep editors and review tools up to date for security reasons.

> https://gist.github.com/lthls/95e029513d6135c21c9fff09bcd4a3e2

Interesting here is that unlike in the gist from @dbuenzli this doesn't show a warning from Github. Also interestingly on macOS I have a hard time to find an editor or other tool which shows the code in a different way - only hexdump and good old vi did the job:
```
let access_level = "user"
let main () =
  if access_level <> "user<200b>" then (* Check if admin *)
    Printf.printf "You are an admin.\n"

let () = main ()
```

**Comment by dbuenzli** [comment]
> If we have a clear criterion and an easy way of implementing it, I'm personally in favor of a new warning (enabled by default).

I'm not against that either but by default the encoding of OCaml sources remains latin1. UTF-8 works "by chance". 

So before having a default warning that checks things related to Unicode in the sources that's the first thing that would need to change in some way, long discussion [here](https://github.com/ocaml/ocaml/pull/1802) (if you ask me I would break compat. mandate *and check* UTF-8 validity and retain latin1  behind a flag for the people who still have code with accented identifier, but that's not a trivial amount of work). Otherwise the compiler view on source encodings becomes a bit non sensical. 

Also the problem is likely a bit more involved than just check the literals, since you can do funky stuff with the comments. The problem with comments is that they are also used to generate documentation and you could then find legitimate uses of these characters which at the moment you cannot escape like in string literals.

**Comment by xavierleroy** [comment]
> It is new (to me) that one has to keep editors and review tools up to date for security reasons.

All software on all your computers should be kept up to date for security reasons.   A bug in a text editor can create havoc too.

There are many ways to use Unicode for obfuscating a code change, beyond the use of bidirectional markers documented in the "Trojan source" paper.   We need a more comprehensive analysis before we can discuss countermeasures in the compiler or elsewhere.

**Comment by xavierleroy** [comment]
I agree with @dbuenzli that we also need to make OCaml "UTF-8 clean" before considering additional safety checks on input files.  I'm optimistic we can find a way forward that still accepts more than US-ASCII inside identifiers, but it's not going to happen overnight either.

**Comment by Octachron** [comment]
Python has an analysis of unicode-related security issues which seems interesting: https://www.python.org/dev/peps/pep-0672 .
I am particularly fond of :
```python
s = "א" * 100 #    "א" is assigned
```

**Comment by xavierleroy** [comment]
Ironically, your URL contains mysterious invisible characters :-)  https://www.python.org/dev/peps/pep-0672/

**Comment by dbuenzli** [comment]
@xavierleroy note that UTF-8 cleanness doesn't have to be tied with allowing Unicode identifiers.

Also I'm personally totally unaware of what happens to these latin1 identifiers in linkers etc. I was just wondering whether the following could be an alternate way forward (rather than what I suggested in my previous message). Since latin1 is fully injected into Unicode (the precise map can be found [here](https://www.unicode.org/Public/MAPPINGS/ISO8859/8859-1.TXT), spoiler: it's the identity function). We could:

1. Mandate sources to be UTF-8 encoded.
2. Comments and literals accept any Unicode character.
2. Identifiers can be US-ASCII or Unicode latin1 characters (we can then even store them as latin1 if the compiler pipeline relies on this).

This still breaks compat. but rather than introducing a new flag the measure is now simply for people to reencode their files from latin1 to UTF-8 and your sources with accented characters will still work. 

The problem of allowing more Unicode characters can be left for later and this way is likely much simpler implementation wise than trying to have two lexing modes.

**Comment by Octachron** [comment]
> Ironically, your URL contains mysterious invisible characters :-) https://www.python.org/dev/peps/pep-0672/

At least, now I know that github doesn't handle gracefully non-breaking spaces in comments.

**Comment by xavierleroy** [comment]
[ This is getting off-topic for this issue report, but for the record: ]

> This still breaks compat. but rather than introducing a new flag the measure is now simply for people to reencode their files from latin1 to UTF-8 and your sources with accented characters will still work.

Right, that's pretty much what I had in mind.  The compiler back-end doesn't care which character encoding is used, and can handle arbitrary bytes in identifiers (or so I think).  So, it's mainly a problem of 1- lexical analysis of identifiers, and 2- normalizing their internal representation (e.g. NFC or NFD).

**Comment by dbuenzli** [comment]
Sorry for the off-topic. 

> 2- normalizing their internal representation (e.g. NFC or NFD).

If you want to make the simplest possible step, I don't think you even need to bring normalization in. With the idea of going towards UTF-8 without leaving previous latin1 users in the dust, you can simply implicitly ask for files to be NFC encoded. That's what usually happens with text input and editors anyways and I'm pretty sure that's what `iconv`  would do if you convert from latin1 to UTF-8. In other words you say files are UTF-8 encoded and replace this sentence of the manual on [identifiers](https://ocaml.org/manual/lex.html#sss:lex:identifiers): 

> The current implementation also recognizes as letters some characters from the ISO 8859-1 set (characters 192–214 and 216–222 as uppercase letters; characters 223–246 and 248–255 as lowercase letters).

By: 

> The current implementation also recognizes as letter the following characters from Unicode (characters U+00C0-U+00D6 and U+00D8-U+00DE as uppercase letters; characters U+00DF-U+00F6 and U+08CF-U+00FF as lowercase letters.

**Comment by dbuenzli** [comment]
Going back on the bidi topic. 

As mentioned in the paper in §VII.F a nicer answer to the problem is to make sure that bidi control characters are not allowed to  "overflow" the programming language's literal and comment delimiters. 

In other words one need to make sure the source character stream forces the bidi algorithm to treat string literals and comments as isolated paragraphs (the level at which the [bidi algorithm](https://www.unicode.org/reports/tr9/) works) and raise a syntax error otherwise. 

For example:
```
"user<RLO><LRI> (* Check if admin *)<PDI><LRI>" 
```
would be illegal because you have an unbounded RLO (right-to-left override) and LRI (left-to-right isolate) in the string literal.

However the following string where RLO and PDI are properly balanced with matching PDI (pop directional isolate) and PDF (pop direcitonal formatting) would be legal: 

```
"user<RLO><LRI> (* Check if admin *)<PDI><LRI><PDI><PDF>"
```
And people who really need literals with unpaired bidi control characters can use escape sequences.

Just a new kind of parentheses to balance somehow :-) 


---
Also I still haven't fully wrapped my head around UAX9 but I also suspect that compiler enforcing the presence of LRI/PDI around string literals once it is found to have characters with directionality which would start influencing the text beyond the string literal delimiter (which is bidi neutral) would be able to solve the surprising python examples.

**Comment by xavierleroy** [comment]
> In other words one need to make sure the source character stream forces the bidi algorithm to treat string literals and comments as isolated paragraphs (the level at which the bidi algorithm works) and raise a syntax error otherwise.

If you can turn this idea into an algorithm, this would be very interesting indeed.  There's room for nice formal verification work here, proving (or disproving) that such an algorithm prevents visual confusion.

**Comment by MSoegtropIMC** [comment]
Github reminded us that it has been a year since we last discussed this.

I am still in favour of the solution suggested by @dbuenzli on Nov 7 2021 (make sure bidirectional formatting cannot have an effect outside of string literals).

My question: is there a consensus on this and the question is only who does it, or are there objections against implementing this suggestion (assuming one can find a reasonable algorithm to decide this, which I expect).

**Comment by gasche** [comment]
Yes, I think that there is a consensus, but no one has offered yet to work on it.

**Comment by xavierleroy** [comment]
I'll offer a dissenting opinion: this "bidi attack" would be better handled by tools that display source code (editors, IDEs, Github/Gitlab) than by compilers.  I think Github already warns in this case.

**Comment by MSoegtropIMC** [comment]
The downside of handling this on the editor side is that for one given language there are many editors / code viewers, but only one compiler. And different people have different preferences here. So while in a given SW project one has detailed control over the compiler used, one has very little control over the editing tools used. Also I am quite sure that forcing people to use certain editors will lead to unpleasant scenes.

One might have control over code viewers used for review in a project, but I personally usually review complex code changes in my favourite editor and just tick it off in the review tool (after checking that the git hash matches). And many editors don't display any warning here (e.g. VSCode).

In the end one should have both, but IMHO checks in the compiler are stronger.

**Comment by MSoegtropIMC** [comment]
P.S.: I just tested again the sequence posted in this [comment](https://github.com/ocaml/ocaml/issues/10749#issuecomment-961982087) in VSCode - no warning

**Comment by dbuenzli** [comment]
Personally I still think that this is a very well marketed POC – which given where it comes from is not surprising, people are keen to issue press releases there. However turning it into a real attack vector is likely to prove quite challenging. 

There are quite a few tools, starting with terminals which are likely to show you the emperor naked. If I wanted to introduce vulnerabilities in code without blowing my cover, that's not the way I would go about it.

In any case if you are particularly paranoid or working on security sensitive code, what I mentioned above, i.e.:

> But the problem with warnings is that they can also be disabled and from the sources itself. I wonder if in the end these things would not be better checked by some form of external dedicated security linter tool.

is likely a much better option. 

Other than I just want to mention that on the Unicode side some kind of working group was formed to work on this issue. Details in the pdf available [from this page](https://home.unicode.org/avoiding-source-code-spoofing/). So we may get extraordinarily complex recommendations to implement in a few years :-)

**Comment by MSoegtropIMC** [comment]
@dbuenzli : thanks for the Unicode working group link - the PDF linked from there is interesting.

Since the PR linked by @DemiMarie would open another can of worms (confusables), I guess it would be best to wait until the OCaml + UTF-8 question is decided and the Unicode experts come up with a recommendation.

Indeed meanwhile I can just use a simple script to check this as long as I don't need any of this in the source code (this is much easier to decide for a concrete project than for a language).

Shall we leave this open as a reminder or close?

**Comment by dbuenzli** [comment]
> Since the PR linked by @DemiMarie would open another can of worms (confusables)

It doesn't if [this simple plan](https://github.com/ocaml/ocaml/issues/10749#issuecomment-962102708) is followed.

**Comment by MSoegtropIMC** [comment]
The discussion in the PR was quite heated - from the up and down votes I concluded that the consensus tended more towards allowing arbitrary unicode identifiers - not just those available in Latin-1. But it is definitely hard to tell what the outcome of this discussion is.

**Comment by dbuenzli** [comment]
Well there's pies in the sky and what the PR actually does which is to enforce that identifiers are US-ASCII only and sources UTF-8 encoded (which enables UTF-8 encoded string literals and comments). 

So the PR doesn't open any can of worms, let's not make it look at more complicated than it is.

**Comment by xavierleroy** [comment]
>  for one given language there are many editors / code viewers, but only one compiler

In my life as a programmer, there are only two editors / code viewers (Emacs and Github), but many compilers: OCaml, GCC, Clang, CompCert, LaTeX, various scripting languages...  Handling "bidi" attacks in every compiler feels like a terrible duplication of effort to me.

**Comment by xavierleroy** [comment]
> It doesn't if https://github.com/ocaml/ocaml/issues/10749#issuecomment-962102708 is followed.

Since this old topic is coming back, I thought I'd submit my old proof-of-concept as a draft PR: #11717.  Perhaps the discussion can continue there.

**Comment by MSoegtropIMC** [comment]
> In my life as a programmer, there are only two editors / code viewers (Emacs and Github), but many compilers: OCaml, GCC, Clang, CompCert, LaTeX, various scripting languages... Handling "bidi" attacks in every compiler feels like a terrible duplication of effort to me.

This is the perspective of an individual developer. Please also put yourself into the perspective of a single industrial project with 100s of developers. There the ratio between numbers of compilers and number of editors is quite opposite - and IMHO also the conclusion on what is a duplication of effort.

Please consider which of these perspectives is more relevant for the topic we are discussing here.

**Comment by Octachron** [comment]
I am not sure what is supposed to be the threat model for a industrial project with 100s developer for such unicode confusion attack? 

As far as I understand, the context where the attack may make sense is for an open source project where it could possibly happen in theory under rare circumstance that a patch request might be different of what it appears to be with a clever use of BIDI characters. In that context, reviewing with a editor that offers an option to reject unbounded BIDI effect inside string literal should be enough.

**Comment by gasche** [comment]
Personally my opinion on the topic is simple: there is a spec that makes sense, so if someone was motivated enough to implement it (and the implementation does not seem like a terrible maintenance burden), I would support integrating it in the compiler. I'm not particular asking for someone to do it either.

**Comment by dbuenzli** [comment]
There is now a [proposed UTS](https://www.unicode.org/reports/tr55/tr55-1.html) (didn't have the time to go over it). If you want your voice to be heard you can feedback until the 2023-01-03 [here](https://www.unicode.org/review/pri466/).

**Comment by gasche** [comment]
Note: the [arxiv paper](https://arxiv.org/abs/2111.00169) on Trojan Source has been updated recently, with the following change description:

> To appear in the 32nd USENIX Security Symposium. Revisions: Adds 4 languages, 2 encodings, threat model, & scanning details
